### PR TITLE
Adapt suggestion search order

### PR DESCRIPTION
**Pull request by simon-isler**
- refactorings & tests for current implementation
- adapted order of suggestion result set

### CODE REVIEW

File: app/controllers/completion_controller.rb [hunk 1]
```diff
@@ -1,59 +1,24 @@
...
-  end
-
-  def issue_order(issue_query)
-    issue_query.order(id: :desc)
-  end
-
-  def order_issues(issues)
-    issues.compact.sort_by(&:id).reverse.uniq(&:id)
+  def search_params
+    params.permit(:q, :term)
```
Comment (simon-isler) [inline app/controllers/completion_controller.rb:12]: What is `q` for?

File: app/services/issue_searcher.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  SEARCH_LIMIT = 10
+  PROJECT_RESULTS_LIMIT = 10
+  TICKET_OPEN_STATUS = true
+
+  def call(search_term, scope)
+    issues = if search_term.present?
+               search_by_term(search_term, scope)
+             else
+               filter_closed_issues(scope).limit(SEARCH_LIMIT).order(id: :desc)
+             end
```
Comment (simon-isler) [inline app/services/issue_searcher.rb:13]: Is this case necessary? Wasn't able to trigger it from the UI itself

File: app/services/issue_searcher.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+         .open(TICKET_OPEN_STATUS)
+         .order('time_entries_count DESC')
+         .limit(PROJECT_RESULTS_LIMIT)
+  end
+
+  def hits_by_subject(search_term, scope)
+    filter_closed_issues(scope.like(search_term)).order(id: :desc).limit(SEARCH_LIMIT)
+  end
+
+  def filter_closed_issues(issues)
```
Comment (schmijos) [inline app/services/issue_searcher.rb:48]: I'd remove this and replace all calls with `.open(true)`

File: app/services/issue_searcher.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class IssueSearcher
+  SEARCH_LIMIT = 10
+  PROJECT_RESULTS_LIMIT = 10
+  TICKET_OPEN_STATUS = true
```
Comment (schmijos) [inline app/services/issue_searcher.rb:6]: I'd kill this constant.

File: app/services/issue_searcher.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+             end
+
+    issues.compact.uniq(&:id)
+  end
+
+  private
+
+  def search_by_term(search_term, scope)
+    found_issues = []
+    found_issues << find_by_id(search_term, scope)
```
Comment (schmijos) [inline app/services/issue_searcher.rb:22]: This adds `nil` sometimes.

File: app/services/issue_searcher.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  def find_by_id(search_term, scope)
+    return unless search_term =~ /\A#?(\d+)\z/
+
+    scope.find_by(id: Regexp.last_match(1).to_i)
+  end
+
+  def hits_by_project(search_term)
+    project_arel = Project.arel_table
+    Issue.left_joins(:project, :time_entries)
+         .select('issues.*, COUNT(time_entries.id) as time_entries_count')
```
Comment (schmijos) [inline app/services/issue_searcher.rb:36]: Since we return `id`s at the end, there is no need to select `issues.*`, right?

**Review by schmijos — APPROVED** [review]

File: app/services/issue_searcher.rb (hunk 3 above)
Comment (simon-isler) [reply app/services/issue_searcher.rb:48]: yes, sure

File: app/services/issue_searcher.rb (hunk 5 above)
Comment (simon-isler) [reply app/services/issue_searcher.rb:22]: yes, there is a `compact` call done on the final result set

File: app/services/issue_searcher.rb (hunk 6 above)
Comment (simon-isler) [reply app/services/issue_searcher.rb:36]: we return `Issue` records and not just the ids

**Pull request merged**